### ISSUE TITLE
optim LAMB assign order issue

### ISSUE THREAD

**Issue by chenyuxyz** [geohot/tinygrad#13600]
in Adam we currently update b1_t, b2_t at the start of step, and the output is different if we move the update to the end of the step (with adjustment to init value), and adding intermediate realize fixes the issue which suggests it's something with assign

change init to
```
    self.b1_t, self.b2_t = (Tensor.full((1,), b, dtype=dtypes.float32, device=self.device, requires_grad=False).contiguous() for b in [b1, b2])
```
and move
```
    self.b1_t *= self.b1
    self.b2_t *= self.b2
```
to after the loop, then `python -m pytest test/test_optim.py -k adam` failed

but adding a realize to `ret.append((t.detach() - self.lr * r * up).cast(t.dtype))` then it passed again

**Comment by aaditya-biswas** [comment]
Hi there! I’m new to contributing here, but I’d love to work on this issue and contribute a fix. I’ll make sure to follow the contribution guidelines and share the PR here as soon as it’s ready. Excited to get started!

**Comment by chenyuxyz** [comment]
fixed #14754
